packages/compass-crud/src/components/readonly-document.tsx

```diff
@@ -30,10 +31,73 @@ export type ReadonlyDocumentProps = {
   showInsights?: boolean;
 };
 
+type ReadonlyDocumentState = {
+  expanded: boolean;
+};
+
 /**
  * Component for a single readonly document in a list of documents.
  */
-class ReadonlyDocument extends React.Component<ReadonlyDocumentProps> {
+class ReadonlyDocument extends React.Component<
+  ReadonlyDocumentProps,
+  ReadonlyDocumentState
+> {
+  constructor(props: ReadonlyDocumentProps) {
+    super(props);
+    this.state = {
+      expanded: props.doc.expanded,
+    };
+  }
+
+  /**
+   * Subscribe to the update store on mount.
+   */
+  componentDidMount() {
+    this.subscribeToDocumentEvents(this.props.doc);
+  }
+
+  /**
+   * Refreshing the list updates the doc in the props so we should update the
+   * document on the instance.
+   */
+  componentDidUpdate(prevProps: ReadonlyDocumentProps) {
+    if (prevProps.doc !== this.props.doc) {
+      this.unsubscribeFromDocumentEvents(prevProps.doc);
+      this.subscribeToDocumentEvents(this.props.doc);
+    }
+  }
+
+  /**
+   * Unsubscribe from the update store on unmount.
+   */
+  componentWillUnmount() {
+    this.unsubscribeFromDocumentEvents(this.props.doc);
+  }
+
+  /**
+   * Subscribe to the document events.
+   */
+  subscribeToDocumentEvents(doc: Document) {
+    doc.on(DocumentEvents.Expanded, this.handleExpanded);
+    doc.on(DocumentEvents.Collapsed, this.handleCollapsed);
+  }
+
+  /**
+   * Unsubscribe from the document events.
+   */
+  unsubscribeFromDocumentEvents(doc: Document) {
+    doc.on(DocumentEvents.Expanded, this.handleExpanded);
+    doc.on(DocumentEvents.Collapsed, this.handleCollapsed);
+  }
+
+  handleExpanded = () => {
+    this.setState({ expanded: true });
+  };
+
+  handleCollapsed = () => {
+    this.setState({ expanded: false });
+  };
```

Comment:
This is mostly copied from the `EditableDocument` to react to relevant events on the `doc`.